### ISSUE TITLE
SSL connection still fails when validatesSecureCertificate = NO

### ISSUE THREAD

**Issue by danthorpe** [pokeb/asi-http-request#134]
I have an iOS application which is using ASIHTTPRequest v1.8-33 to make HTTPS calls to a server. The secure connection has a certificate, but it's provided by my hosting company, so it has a mismatched domain. The Certificate is for "*.acme-hosting-company.com" and I'm accessing "secure.mydomain.com".

My Objective-C code is basically:

```
// Set credentials
[request setValidatesSecureCertificate:NO];
if (self.login && self.password) {
    request.username = self.login;
    request.password = self.password;
}

[request startSynchronous];
NSError *error = [request error];
if (error) {
    NSLog(@"Error: %@", [error userInfo]);
    NSLog(@"response string: %@",[request responseString]); 
    return nil;
}
```

Which fails with the messages:

```
response string: (null)
Error: {
  NSLocalizedDescription = "A connection failure occurred: SSL problem (possibly a bad/expired/self-signed certificate)";
  NSUnderlyingError = "Error Domain=NSOSStatusErrorDomain Code=-9806 \"The operation couldn't be completed. (OSStatus error -9806.)\" UserInfo=0x2f0d10 {}";
}
```

So, until I get a proper certificate, it was my understanding that setting validatesSecureCertificate to false would take care of this? The certificate is signed by a trusted party, but it's for a mismatched domain, so I'm guessing this is what's causing it to fail?

Any thoughts?

Cheers,
Dan

And thanks for the framework - it's very easy to use!

**Comment by pokeb** [comment]
I think the error message may be slightly misleading - according to SecureTransport.h, -9806 is 'connection closed via error'. What happens if you connect in Safari - does it just warn you about the domain mismatch, and let you connect?

**Comment by danthorpe** [comment]
That's right, it pops up a warning about the domain mismatch and let you proceed.

**Comment by pokeb** [comment]
Can you share the real url? Msg me or send me an email (ben@allseeing-i.com) if you don't want to post it here.

**Comment by danthorpe** [comment]
Okay, I set up a new domain for testing purposes - https://ssltest.blindingskies.com. It just contains "Hello, World". If you go there in Safari it will give you a warning, with the option to Show Certificate, Cancel or Continue. If you click continue, I don't think it adds the certificate to the Keychain, but equally, it doesn't prompt you again, so presumably Safari/CFNetwork caches it somewhere. If you click Show Certificate, then there is an option to install it in the keychain.

Unfortunately, I can't really re-create my exact test from before, as I've not got a proper certificate installed on my domain, and it's quite a complex project that I don't want to dick around with. But you can play about with ssltest as long as you want.

**Comment by pokeb** [comment]
In quick test it seems to work for me:

```
ASIHTTPRequest *request = [ASIHTTPRequest requestWithURL:[NSURL URLWithString:@"https://ssltest.blindingskies.com"]];
[request setValidatesSecureCertificate:NO];
[request setUsername:@"Ben"];
[request setPassword:@"Copsey"];
[request startSynchronous];
if ([request error]) {
    NSLog(@"%@",[[request error] localizedDescription]);
} else {
    NSLog(@"%@",[request responseString]);
}
```

Does this code work for you? If you can modify the above to demonstrate the issue, that would be helpful.

**Comment by danthorpe** [comment]
Hmm, I've not had a chance yet to test your code, but, I guess my previous error might have been related to an authentication problem. At the original domain, I was running a CouchDB instance that required a valid user. The ssltest domain above is just dumb static html, without any authentication required. I'll see if can get a CouchDB up there to test.
